sub/internal/storage/subscriber/integration_test.go

```diff
@@ -0,0 +1,222 @@
+//go:build integration
+
+package subscriber
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/stretchr/testify/require"
+
+	"github.com/GenesisEducationKyiv/software-engineering-school-4-0-hrvadl/sub/internal/storage/platform/db"
+)
+
+const testDSNEnvKey = "SUB_TEST_DSN"
+
+func TestMain(t *testing.M) {
+	code := t.Run()
+	dsn := os.Getenv(testDSNEnvKey)
+
+	db, err := db.NewConn(dsn)
+	if err != nil {
+		panic("failed to connect to test db")
+	}
+
+	if _, err := db.Exec("DELETE FROM subscribers"); err != nil {
+		panic("failed to cleanup")
+	}
+
+	os.Exit(code)
+}
+
+func TestSave(t *testing.T) {
+	type args struct {
+		ctx context.Context
+		sub Subscriber
+	}
+	testCases := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "Should save subscriber correctly",
+			args: args{
+				ctx: context.Background(),
+				sub: Subscriber{Email: "test1@mail.com"},
+			},
+			wantErr: false,
+		},
+		{
+			name: "Should not get subscribers correctly when it takes too long",
+			args: args{
+				ctx: newImmediateCtx(),
+			},
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			dsn := os.Getenv(testDSNEnvKey)
+			require.NotZero(t, dsn, "test DSN can not be empty")
+			db, err := db.NewConn(dsn)
+			require.NoError(t, err, "Failed to connect to test DB")
+
+			r := NewRepo(db)
+			id, err := r.Save(tt.args.ctx, tt.args.sub)
+			t.Cleanup(func() {
+				cleanupSub(t, db, id)
+			})
+
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+
+			require.NoError(t, err)
+			require.NotZero(t, id)
+		})
+	}
+}
+
+func TestSaveSubscriberTwice(t *testing.T) {
+	type args struct {
+		ctx context.Context
+		sub Subscriber
+	}
+	testCases := []struct {
+		name string
+		args args
+	}{
+		{
+			name: "Should not save subscriber twice",
+			args: args{
+				ctx: context.Background(),
+				sub: Subscriber{Email: "test1@mail.com"},
+			},
+		},
+		{
+			name: "Should not save subscriber twice",
+			args: args{
+				ctx: context.Background(),
+				sub: Subscriber{Email: "test@mail.com"},
+			},
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			dsn := os.Getenv(testDSNEnvKey)
+			require.NotZero(t, dsn, "test DSN can not be empty")
+			db, err := db.NewConn(dsn)
+			require.NoError(t, err, "Failed to connect to test DB")
+
+			r := NewRepo(db)
+			id, err := r.Save(tt.args.ctx, tt.args.sub)
+			t.Cleanup(func() {
+				cleanupSub(t, db, id)
+			})
+
+			require.NoError(t, err)
+			require.NotZero(t, id)
+
+			id, err = r.Save(tt.args.ctx, tt.args.sub)
+			require.Error(t, err)
+			require.Zero(t, id)
+		})
+	}
+}
+
+func TestGet(t *testing.T) {
```

Comment:
while shorter test names are preferred, imo too ambiguous what "get" means and can be understood only after reading the code or the context of the whole file

Reply:
ok, you're right